### PR TITLE
HDDS-8095. Unbuffer not supported in TDE buckets

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?

Add StreamCapabilities interface to ExtendedInputStream to allow applications to unbuffer input streams in encrypted buckets.

Because StreamCapabilities is a Hadoop API that does not exist in 2.7, this change would have broken applications if Hadoop2 runtime is used. This PR employs a hack which copies StreamCapabilities as well as StreamCapabilitiesPolicy interfaces to the ozonefs-hadoop2 module so application wouldn't fail to find the class definition at runtime.

A unit test is added to verify that unbuffer() invoked on a CryptoInputStream does invoke the unbuffer() of the wrapped KeyInputStream.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8095

## How was this patch tested?

Unit test.

### CODE REVIEW

File: hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/package-info.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Utility classes copied from Hadoop3 to maintain Hadoop2 runtime
+ * compatibility.
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Evolving
+package org.apache.hadoop.fs;
+
+import org.apache.hadoop.hdds.annotation.InterfaceAudience;
+import org.apache.hadoop.hdds.annotation.InterfaceStability;
```
Comment (Galsza) [inline hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/package-info.java:29]: No new line at the end of file. We should have a checkstyle rule for this point, I'm not sure why this passed.

File: hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/StreamCapabilities.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * Stream read(long, ByteBuffer) capability implemented by
+   * {@link ByteBufferPositionedReadable#read(long, java.nio.ByteBuffer)}.
+   */
+  String PREADBYTEBUFFER = "in:preadbytebuffer";
+
+  /**
+   * IOStatisticsSource API.
+   */
+  //String IOSTATISTICS = "iostatistics";
```
Comment (Galsza) [inline hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/StreamCapabilities.java:86]: Is this supposed to be commented out?

File: hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/StreamCapabilities.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * Stream abort() capability implemented by {@link Abortable#abort()}.
+   * This matches the Path Capability
+   * {@link CommonPathCapabilities#ABORTABLE_STREAM}.
+   */
+  //String ABORTABLE_STREAM =  CommonPathCapabilities.ABORTABLE_STREAM;
```
Comment (Galsza) [inline hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/StreamCapabilities.java:93]: Same here, is this supposed to be commented out?

**Comment by Galsza** [comment]
Thanks for working on this @jojochuang 

The change overall looks good.

Can I ask for learning purposes how is `StreamCapabilitiesPolicy` picked up by the framework? I don't see a class extending it or it overriding the unbuffer method.

File: hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/package-info.java (hunk 1 above)
Comment (jojochuang) [reply hadoop-ozone/ozonefs-hadoop2/src/main/java/org/apache/hadoop/fs/package-info.java:29]: Weird -- I did rebase onto the latest master but somehow it's not being picked up. Will update the PR.

**Comment by jojochuang** [comment]
StreamCapabilitiesPolicy is used by Hadoop's CryptoInputStream.unbuffer(). So that's why I included in the first place. But if an installation requires Hadoop 2 dependency, They probably shouldn't invoke unbuffer API any way (it's not part of Hadoop 2.7 API). So I'll drop StreamCapabilitiesPolicy from the PR.

**Comment by jojochuang** [comment]
The fix is verified by Impala.

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @jojochuang for the patch, @Galsza for the review.

**Pull request merged**